### ISSUE TITLE
[Regression] Conda create environment fails on lock if root environment is not under user control

### ISSUE THREAD

**Issue by frol** [conda/conda#3307]
This is "funny", but it seems that Conda managed to break this thing the second time in a month... #2681 was the previous one.

This time, I get the following error:

```
$ conda create -n _root --yes --use-index-cache python=3 
...
    Traceback (most recent call last):
      File "/usr/local/miniconda/lib/python3.5/site-packages/conda/exceptions.py", line 442, in conda_exception_handler
        return_value = func(*args, **kwargs)
      File "/usr/local/miniconda/lib/python3.5/site-packages/conda/cli/main.py", line 144, in _main
        exit_code = args.func(args, p)
      File "/usr/local/miniconda/lib/python3.5/site-packages/conda/cli/main_create.py", line 66, in execute
        install(args, parser, 'create')
      File "/usr/local/miniconda/lib/python3.5/site-packages/conda/cli/install.py", line 399, in install
        execute_actions(actions, index, verbose=not args.quiet)
      File "/usr/local/miniconda/lib/python3.5/site-packages/conda/plan.py", line 640, in execute_actions
        inst.execute_instructions(plan, index, verbose)
      File "/usr/local/miniconda/lib/python3.5/site-packages/conda/instructions.py", line 132, in execute_instructions
        cmd(state, arg)
      File "/usr/local/miniconda/lib/python3.5/site-packages/conda/instructions.py", line 77, in LINK_CMD
        link(state['prefix'], dist, lt, index=state['index'])
      File "/usr/local/miniconda/lib/python3.5/site-packages/conda/install.py", line 1060, in link
        with DirectoryLock(prefix), FileLock(source_dir):
      File "/usr/local/miniconda/lib/python3.5/site-packages/conda/lock.py", line 86, in __enter__
        touch(self.lock_file_path)
      File "/usr/local/miniconda/lib/python3.5/site-packages/conda/lock.py", line 48, in touch
        with open(file_name, 'a'):
    PermissionError: [Errno 13] Permission denied: '/usr/local/miniconda/pkgs/openssl-1.0.2h-1.pid34.conda_lock'
```

```
Current conda install:

               platform : linux-64
          conda version : 4.2.2
       conda is private : False
      conda-env version : 2.6.0
    conda-build version : 1.21.11+0.g5b44ab3.dirty
         python version : 3.5.2.final.0
       requests version : 2.10.0
       root environment : /usr/local/miniconda  (read only)
    default environment : /usr/local/miniconda
       envs directories : /home/gitlab-ci/.conda/envs
                          /usr/local/miniconda/envs
          package cache : /home/gitlab-ci/.conda/envs/.pkgs
                          /usr/local/miniconda/pkgs
           channel URLs : defaults
            config file : None
           offline mode : False
```

/CC @kalefranz

**Comment by kalefranz** [comment]
@frol, First, **thanks for using canary**!!  Do you know if a previous run of conda crashed, exited prematurely, or you sent a signal (i.e. ctrl-c) for it to exit early?  Something made that lock file stick around...

    conda clean --yes --lock

should take care of it for you.  If it doesn't, that's definitely a bug.

We're still working on tuning the locking in conda.  It's (hopefully) better in 4.2 than it was in 4.1, and I know for sure it will be better yet in 4.3 with @HugoTian's work in #3197.

**Comment by frol** [comment]
@kalefranz The error message is quite informative:

```
     File "/usr/local/miniconda/lib/python3.5/site-packages/conda/lock.py", line 48, in touch
        with open(file_name, 'a'):
   PermissionError: [Errno 13] Permission denied: '/usr/local/miniconda/pkgs/openssl-1.0.2h-1.pid34.conda_lock'
```

`/usr/local/miniconda` is a read-only directory, so there is no way `open(mode='a')` will ever succeed there.

**Comment by kalefranz** [comment]
Ahh. This then is what I generally call "multi-user support." That is, install conda as one user, and use the executable as another but without write permissions to the conda executable's root environment. 

This used case has always been problematic, and it's s high priority to tackle and "get right" in the next several months.

**Comment by davidslac** [comment]
I think I just ran into this, created a env in a central install area, then from a user account tried to clone the environment, but it failed with something similar, doing

`conda create --offline -n myrel --clone anarel-1.0.0
`
caused
```
...
WARNING conda.install:warn_failed_remove(178): Cannot remove, permission denied: /reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7/pkgs/psana-conda-1.0.0-py27_1/bin
WARNING conda.install:warn_failed_remove(178): Cannot remove, permission denied: /reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7/pkgs/psana-conda-1.0.0-py27_1
Pruning fetched packages from the cache ...
An unexpected error has occurred.
Please consider posting the following information to the
conda GitHub issue tracker at:

    https://github.com/conda/conda/issues



Current conda install:

               platform : linux-64
          conda version : 4.2.3
       conda is private : False
      conda-env version : 2.6.0
    conda-build version : 1.21.11+0.g5b44ab3.dirty
         python version : 2.7.12.final.0
       requests version : 2.10.0
       root environment : /reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7  (read only)
    default environment : /reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7/envs/anarel-1.0.0
       envs directories : /reg/neh/home/davidsch/.conda/envs
                          /reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7/envs
          package cache : /reg/neh/home/davidsch/.conda/envs/.pkgs
                          /reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7/pkgs
           channel URLs : file:///reg/g/psdm/sw/conda/channels/system-rhel7
                          file:///reg/g/psdm/sw/conda/channels/external-rhel7
                          defaults
                          file:///reg/g/psdm/sw/conda/channels/psana-rhel7
                          scikit-beam
                          file:///reg/g/psdm/sw/conda/channels/testing-rhel7
            config file : /reg/neh/home/davidsch/.condarc
           offline mode : False



`$ /reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7/envs/anarel-1.0.0/bin/conda create --offline -n myrel --clone anarel-1.0.0`




    Traceback (most recent call last):
      File "/reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7/lib/python2.7/site-packages/conda/exceptions.py", line 442, in conda_exception_handler
        return_value = func(*args, **kwargs)
      File "/reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7/lib/python2.7/site-packages/conda/cli/main.py", line 144, in _main
        exit_code = args.func(args, p)
      File "/reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7/lib/python2.7/site-packages/conda/cli/main_create.py", line 66, in execute
        install(args, parser, 'create')
      File "/reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7/lib/python2.7/site-packages/conda/cli/install.py", line 225, in install
        clone(args.clone, prefix, json=args.json, quiet=args.quiet, index_args=index_args)
      File "/reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7/lib/python2.7/site-packages/conda/cli/install.py", line 90, in clone
        index_args=index_args)
      File "/reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7/lib/python2.7/site-packages/conda/misc.py", line 388, in clone_env
        force_extract=False, index_args=index_args)
      File "/reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7/lib/python2.7/site-packages/conda/misc.py", line 188, in explicit
        execute_actions(actions, index=index, verbose=verbose)
      File "/reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7/lib/python2.7/site-packages/conda/plan.py", line 639, in execute_actions
        inst.execute_instructions(plan, index, verbose)
      File "/reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7/lib/python2.7/site-packages/conda/instructions.py", line 132, in execute_instructions
        cmd(state, arg)
      File "/reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7/lib/python2.7/site-packages/conda/instructions.py", line 66, in RM_FETCHED_CMD
        rm_fetched(arg)
      File "/reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7/lib/python2.7/site-packages/conda/install.py", line 801, in rm_fetched
        with FileLock(fname):
      File "/reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7/lib/python2.7/site-packages/conda/lock.py", line 86, in __enter__
        touch(self.lock_file_path)
      File "/reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7/lib/python2.7/site-packages/conda/lock.py", line 48, in touch
        with open(file_name, 'a'):
    IOError: [Errno 13] Permission denied: u'/reg/g/psdm/sw/conda/inst/miniconda2-dev-rhel7/pkgs/openmpi-1.10.3-lsf_verbs_100.tar.bz2.pid17825.conda_lock'

```

**Comment by kalefranz** [comment]
@HugoTian As part of the 4.2 release, I think we need a lock cleanup step in the exception handler that just attempts to clean up any lock files for the current pid before it ultimately exits.  Do you agree?

**Comment by HugoTian** [comment]
Moreover,  I think some how we need to bring back the try/except block when we create the lock file.

Something like : 

             try:
                touch(****)
            except OSError:
                pass

**Comment by HugoTian** [comment]
Please refer to PR #3326.

We may warn user when lock cannot be created , and when lock fails, it won't affect conda .

**Comment by kalefranz** [comment]
fixed with #3326

**Comment by alienzj** [comment]
Nice!
